README.md
# gb.java3

lesson3/History.txt
line 1
line 2
line 3
line 4
line 5
line 6
line 7
line 8
line 9
line 10
line 11
line 12
line 13
line 14
line 15
line 16
line 17
line 18
line 19
line 20
line 21
line 22
line 23
line 24
line 25
line 26
line 27
line 28
line 29
line 30
line 31
line 32
line 33
line 34
line 35
line 36
line 37
line 38
line 39
line 40
line 41
line 42
line 43
line 44
line 45
line 46
line 47
line 48
line 49
line 50
line 51
line 52
line 53
line 54
line 55
line 56
line 57
line 58
line 59
line 60
line 61
line 62
line 63
line 64
line 65
line 66
line 67
line 68
line 69
line 70
line 71
line 72
line 73
line 74
line 75
line 76
line 77
line 78
line 79
line 80
line 81
line 82
line 83
line 84
line 85
line 86
line 87
line 88
line 89
line 90
line 91
line 92
line 93
line 94
line 95
line 96
line 97
line 98
line 99
line 100
line 101
line 102
line 103
line 104
line 105
line 106
line 107
line 108
line 109
line 110
line 111
line 112
line 113
line 114
line 115
line 116
line 117
line 118
line 119
line 120
line 121
line 122
line 123
line 124
line 125
line 126
line 127
line 128
line 129
line 130
[01:18:21] Hunter: Привет
[01:18:25] Hunter: как дела?
[01:18:28] Hunter: что нового?
[01:33:59] Hunter: ghbdtn
[01:34:00] Hunter: aslfj
[01:34:01] Hunter: asf'aks


lesson2/src/main/java/Client2/Client2.java
package Client2;

import Client1.Client;

import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

import java.net.Socket;
import java.text.SimpleDateFormat;
import java.util.Date;


public class Client2 {

    String userName = "Клиент";
    Windows window;

    private final String SERVER_HOST = "localhost";
    private final int SERVER_PORT = 8181;

    private JTextField msgInputField;
    private JTextArea chatArea;

    private Socket socket;
    private DataOutputStream dos;
    private DataInputStream dis;

    private boolean isAuthorized;

    public Client2() {
        try {
            openConnect();
        } catch (IOException e) {
            System.out.println("нет соединения");
        }
        window = new Windows(userName, this);
    }



    private void openConnect() throws IOException {

        socket = new Socket(SERVER_HOST, SERVER_PORT);

        dis = new DataInputStream(socket.getInputStream());
        dos = new DataOutputStream(socket.getOutputStream());
        setAuthorized(false);
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {

                    while (true) {
                        String serverMessege = dis.readUTF();

                        if (serverMessege.startsWith("/")) {
                            if (serverMessege.startsWith("/authok")) {
                                Client2.Windows.text = Client2.Windows.text.concat("Welcome to the CHAT" + "\n");
                                Client2.Windows.setInArea();
                                String[] part = serverMessege.split("\\s");
//                                Arrays.stream(part).forEach(a-> System.out.println(a));
//                                System.out.println(serverMessege);
                                window.windowsName(part[1]);

                                continue;
                            }
                            if (serverMessege.equals("/quit")) {
                                break;
                            }
                            if (serverMessege.startsWith("/renameok")){
                                String[] part = serverMessege.split("\\s");
                                window.windowsName(part[1]);
                            }
                        } else {
                            Client2.Windows.text = Client2.Windows.text.concat("[" + Client2.Windows.date() + "]" + serverMessege + "\n");
                            Client2.Windows.setInArea();
                        }
                    }
                } catch (IOException e) {
                    System.out.println("connection closed");
                }
            }
        }).start();

    }

    public void closeConnection() {
        try {
            dis.close();
        } catch (IOException e) {
            e.printStackTrace();

        } catch (NullPointerException ignored) {
        }
        try {
            dos.close();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (NullPointerException ignored) {
        }
        try {
            socket.close();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (NullPointerException ignored) {
        }
    }

    public static void main(String[] args) {

        Client client = new Client();

    }

    public boolean isAuthorized() {
        return isAuthorized;
    }

    public void setAuthorized(boolean authorized) {
        isAuthorized = authorized;
    }

    static class Windows extends JFrame {

        static String text = "";

        static String name;

        static Client2 client;

        static JTextField field = new JTextField();
        static JTextArea area = new JTextArea();
        static JScrollPane scroll = new JScrollPane(area);


        public static void setInArea() {
            area.setText("" + text);
        }

        public void windowsName(String name){
            setTitle(name);
        }

        public Windows(String name, Client2 client) {
            this.client = client;
            this.name = name;

            area.setEditable(false);
            area.setLineWrap(true);
            windowsName("Please LogIn");
            setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
            setBounds(200, 200, 600, 600);

            //JScrollPane jscroll = new JScrollPane(JScrollPane.VERTICAL_SCROLLBAR_ALWAYS, JScrollPane.HORIZONTAL_SCROLLBAR_AS_NEEDED);
            // jscroll.setSize(new Dimension(20,100));

            JButton btnSend = new JButton("Sent");

            field.addKeyListener(new KeyListener() {
                @Override
                public void keyTyped(KeyEvent e) {
                }

                @Override
                public void keyPressed(KeyEvent e) {
                    if (e.getExtendedKeyCode() == 10 && !(field.getText().trim().equals(""))) {
                        text = text.concat(" " + name + " " + date() + ": " + field.getText() + "\n");
                        area.setText("" + text);
                        sendMessege(field.getText());
                        field.setText("");
                    } else if (e.getExtendedKeyCode() == 10 ){
                        field.setText("");
                    }
                }

                @Override
                public void keyReleased(KeyEvent e) {
                }

            });
            btnSend.addMouseListener(new MouseAdapter() {
                @Override
                public void mouseClicked(MouseEvent e) {
                    if (e.getClickCount() == 1 && !(field.getText().trim().equals(""))) {

                        text = text.concat(" " + name + " " + date() + ": " + field.getText() + "\n");

                        area.setText("" + text);

                        sendMessege(field.getText());

                        field.setText("");


                    }  else if (e.getClickCount() == 1 ) {
                        field.setText("");
                    }
                }
            });

            JPanel upPanel = new JPanel();
            JPanel sendPanel = new JPanel();

            upPanel.setSize(250, 250);


            upPanel.setLayout(new BorderLayout());
            sendPanel.setLayout(new BorderLayout());
            //area.add(jscroll);
//        textField.setSize(1500,500);
            btnSend.setSize(50, 20);

            sendPanel.add(field, BorderLayout.CENTER);
            sendPanel.add(btnSend, BorderLayout.EAST);


            upPanel.add(sendPanel, BorderLayout.SOUTH);
            //       upPanel.add(btn5, BorderLayout.WEST);
//        upPanel.add(jscroll, BorderLayout.EAST);
//       upPanel.add(btn4, BorderLayout.NORTH);
            upPanel.add(scroll, BorderLayout.CENTER);

            addWindowListener(new WindowAdapter() {
                @Override
                public void windowClosing(WindowEvent e) {
                    super.windowClosing(e);
                    try {
                        client.dos.writeUTF("/quit");

                    } catch (IOException ignored) {
                    } catch (NullPointerException ignored) {
                    } finally {
                        client.closeConnection();
                    }
                }
            });


            add(upPanel);
            setVisible(true);

        }

        private void sendMessege(String text) {
            try {
                client.dos.writeUTF(text);
            } catch (Exception e) {
                System.out.println("Сообщение не отправлено");
            }
        }


        private static String date() {

            Date date = new Date();

            SimpleDateFormat ft =
                    new SimpleDateFormat("hh:mm:ss");

            return ft.format(date);


        }

    }

}


lesson2/src/main/java/Client3/Client3.java
package Client3;

import Client1.Client;

import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

import java.net.Socket;
import java.text.SimpleDateFormat;
import java.util.Date;


public class Client3 {

    String userName = "Клиент";
    Windows window;

    private final String SERVER_HOST = "localhost";
    private final int SERVER_PORT = 8181;

    private JTextField msgInputField;
    private JTextArea chatArea;

    private Socket socket;
    private DataOutputStream dos;
    private DataInputStream dis;

    private boolean isAuthorized;

    public Client3() {
        try {
            openConnect();
        } catch (IOException e) {
            System.out.println("нет соединения");
        }
        window = new Windows(userName, this);
    }



    private void openConnect() throws IOException {

        socket = new Socket(SERVER_HOST, SERVER_PORT);

        dis = new DataInputStream(socket.getInputStream());
        dos = new DataOutputStream(socket.getOutputStream());
        setAuthorized(false);
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {

                    while (true) {
                        String serverMessege = dis.readUTF();

                        if (serverMessege.startsWith("/")) {
                            if (serverMessege.startsWith("/authok")) {
                                Client3.Windows.text = Client3.Windows.text.concat("Welcome to the CHAT" + "\n");
                                Client3.Windows.setInArea();
                                String[] part = serverMessege.split("\\s");
//                                Arrays.stream(part).forEach(a-> System.out.println(a));
//                                System.out.println(serverMessege);
                                window.windowsName(part[1]);

                                continue;
                            }
                            if (serverMessege.equals("/quit")) {
                                break;
                            }
                            if (serverMessege.startsWith("/renameok")){
                                String[] part = serverMessege.split("\\s");
                                window.windowsName(part[1]);
                            }
                        } else {
                            Client3.Windows.text = Client3.Windows.text.concat("[" + Client3.Windows.date() + "]" + serverMessege + "\n");
                            Client3.Windows.setInArea();
                        }
                    }
                } catch (IOException e) {
                    System.out.println("connection closed");
                }
            }
        }).start();
    }

    public void closeConnection() {
        try {
            dis.close();
        } catch (IOException e) {
            e.printStackTrace();

        } catch (NullPointerException ignored) {
        }
        try {
            dos.close();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (NullPointerException ignored) {
        }
        try {
            socket.close();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (NullPointerException ignored) {
        }
    }

    public static void main(String[] args) {

        Client client = new Client();

    }

    public boolean isAuthorized() {
        return isAuthorized;
    }

    public void setAuthorized(boolean authorized) {
        isAuthorized = authorized;
    }

    static class Windows extends JFrame {

        static String text = "";

        static String name;

        static Client3 client;

        static JTextField field = new JTextField();
        static JTextArea area = new JTextArea();
        static JScrollPane scroll = new JScrollPane(area);


        public static void setInArea() {
            area.setText("" + text);
        }

        public void windowsName(String name){
            setTitle(name);
        }

        public Windows(String name, Client3 client) {
            this.client = client;
            this.name = name;

            area.setEditable(false);
            area.setLineWrap(true);
            windowsName("Please LogIn");


            setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
            setBounds(200, 200, 600, 600);

            //JScrollPane jscroll = new JScrollPane(JScrollPane.VERTICAL_SCROLLBAR_ALWAYS, JScrollPane.HORIZONTAL_SCROLLBAR_AS_NEEDED);
            // jscroll.setSize(new Dimension(20,100));

            JButton btnSend = new JButton("Sent");

            field.addKeyListener(new KeyListener() {
                @Override
                public void keyTyped(KeyEvent e) {
                }

                @Override
                public void keyPressed(KeyEvent e) {
                    if (e.getExtendedKeyCode() == 10 && !(field.getText().trim().equals(""))) {
                        text = text.concat(" " + name + " " + date() + ": " + field.getText() + "\n");
                        area.setText("" + text);
                        sendMessege(field.getText());
                        field.setText("");
                    } else if (e.getExtendedKeyCode() == 10 ){
                        field.setText("");
                    }
                }

                @Override
                public void keyReleased(KeyEvent e) {
                }

            });
            btnSend.addMouseListener(new MouseAdapter() {
                @Override
                public void mouseClicked(MouseEvent e) {
                    if (e.getClickCount() == 1 && !(field.getText().trim().equals(""))) {

                        text = text.concat(" " + name + " " + date() + ": " + field.getText() + "\n");

                        area.setText("" + text);

                        sendMessege(field.getText());

                        field.setText("");


                    }  else if (e.getClickCount() == 1 ) {
                        field.setText("");
                    }
                }
            });

            JPanel upPanel = new JPanel();
            JPanel sendPanel = new JPanel();

            upPanel.setSize(250, 250);


            upPanel.setLayout(new BorderLayout());
            sendPanel.setLayout(new BorderLayout());
            //area.add(jscroll);
//        textField.setSize(1500,500);
            btnSend.setSize(50, 20);

            sendPanel.add(field, BorderLayout.CENTER);
            sendPanel.add(btnSend, BorderLayout.EAST);


            upPanel.add(sendPanel, BorderLayout.SOUTH);
            //       upPanel.add(btn5, BorderLayout.WEST);
//        upPanel.add(jscroll, BorderLayout.EAST);
//       upPanel.add(btn4, BorderLayout.NORTH);
            upPanel.add(scroll, BorderLayout.CENTER);

            addWindowListener(new WindowAdapter() {
                @Override
                public void windowClosing(WindowEvent e) {
                    super.windowClosing(e);
                    try {
                        client.dos.writeUTF("/quit");

                    } catch (IOException ignored) {
                    } catch (NullPointerException ignored) {
                    } finally {
                        client.closeConnection();
                    }
                }
            });


            add(upPanel);
            setVisible(true);

        }

        private void sendMessege(String text) {
            try {
                client.dos.writeUTF(text);
            } catch (Exception e) {
                System.out.println("Сообщение не отправлено");
            }
        }


        private static String date() {

            Date date = new Date();

            SimpleDateFormat ft =
                    new SimpleDateFormat("hh:mm:ss");

            return ft.format(date);


        }

    }

}


lesson3/src/main/java/Server/MyServer.java
package Server;

import Server.Interface.AuthService;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class MyServer {

    private final int PORT = 8181;

    private List<ClientHandler> clients;
    private AuthService authService;



    public AuthService getAuthService() {
        return authService;
    }



    public MyServer() {
        try (ServerSocket serverSocket = new ServerSocket(PORT)) {
            try {
                authService = new BaseAuthService() {
                };
            } catch (SQLException throwables) {
                throwables.printStackTrace();
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
            }



            authService.start();
            clients = new ArrayList<>();
            while (true) {
                System.out.println("Server wait to connection");
                Socket socket = serverSocket.accept();
                System.out.println("new client connected");
                new ClientHandler(this, socket);
            }


        } catch (IOException e) {
            System.out.println("server error");
        } finally {
            if (authService != null) {
                authService.stop();
            }
        }
    }

    public void subscribe(ClientHandler c) {
        clients.add(c);
    }

    public void unSubscribe(ClientHandler c) {
        clients.remove(c);
    }

    public synchronized boolean isNickBusy(String nick) {
        for (ClientHandler c : clients) {
            if (c.getNickname().equals(nick)) {
                return true;
            }
        }
        return false;
    }

    public synchronized void broadCastMessage (String message){
        for (ClientHandler c: clients) {
            c.sendMessage(" " + message);
        }
    }
    public synchronized boolean privateMessage(String nickName, String message) {
        for (ClientHandler c: clients) {
            if (c.nickname.equals(nickName)){
                c.sendMessage(message);
                return true;
            }
        }
        return false;
    }


    public void allList(ClientHandler client) {
        for (ClientHandler c: clients) {
            client.sendMessage(c.nickname);
        }
    }
}


lesson3/src/main/java/Client1/Client.java
package Client1;

import Client1.Interfaces.History;

import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.io.*;

import java.net.Socket;
import java.text.SimpleDateFormat;
import java.util.Date;


public class Client implements History {

    String userName = "Клиент";
    static Windows window;

    private final String SERVER_HOST = "localhost";
    private final int SERVER_PORT = 8181;

    private Socket socket;
    private DataOutputStream dos;
    private DataInputStream dis;

    private boolean isAuthorized;

    File history = new File("D:\\1\\History.txt");

    public Client() {
        try {
            openConnect();
        } catch (IOException e) {
            System.out.println("нет соединения");
        }
        window = new Windows(userName, this);
    }


    private void openConnect() throws IOException {

        socket = new Socket(SERVER_HOST, SERVER_PORT);

        dis = new DataInputStream(socket.getInputStream());
        dos = new DataOutputStream(socket.getOutputStream());
        setAuthorized(false);

        new Thread(new Runnable() {
            @Override
            public void run() {
                try {

                    while (true) {
                        String serverMessege = dis.readUTF();

                        if (serverMessege.startsWith("/")) {
                            if (serverMessege.startsWith("/authok")) {
                                Windows.text = Windows.text.concat("Welcome to the CHAT" + "\n");

                                String[] part = serverMessege.split("\\s");
                                window.windowsName(part[1]);

                                loadMessages();
                                Windows.setInArea();
                                Windows.area.setCaretPosition(Windows.area.getDocument().getLength());
                                continue;
                            }
                            if (serverMessege.equals("/quit")) {
                                break;
                            }
                            if (serverMessege.startsWith("/renameok")) {
                                String[] part = serverMessege.split("\\s");
                                window.windowsName(part[1]);
                            }
                        } else {
                            StringBuilder stringBuilder = new StringBuilder("[" + Windows.date() + "]" + serverMessege + "\n");
                            saveMessages(stringBuilder.toString());
                            Windows.text = Windows.text.concat(stringBuilder.toString());
                            Windows.setInArea();
                            Windows.area.setCaretPosition(Windows.area.getDocument().getLength());
                        }
                    }
                } catch (IOException e) {
                    System.out.println("connection closed");
                }
            }
        }).start();

    }

    public void closeConnection() {
        try {
            dis.close();
        } catch (IOException e) {
            e.printStackTrace();

        } catch (NullPointerException ignored) {
        }
        try {
            dos.close();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (NullPointerException ignored) {
        }
        try {
            socket.close();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (NullPointerException ignored) {
        }
    }

    public static void main(String[] args) {

        Client client = new Client();
        //window.sendMessege("/auth A A");


    }


    public void setAuthorized(boolean authorized) {
        isAuthorized = authorized;
    }

    @Override
    public void saveMessages(String st) {
        if (st != null) {

            if (!history.exists() || history.isDirectory()) {
                try {
                    history.createNewFile();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            if (history.isFile() && history.canWrite()) {

                try (BufferedWriter writer = new BufferedWriter(new FileWriter(history, true))) {

                    writer.write(st);

                } catch (FileNotFoundException e) {
                    e.printStackTrace();
                } catch (IOException e) {
                    e.printStackTrace();
                }


            }
        }
    }

    @Override
    public void loadMessages() throws IOException {
        String st = "";
        int lines = 0;
        if (history.exists() && history.canRead()) {
            BufferedReader count = new BufferedReader(new FileReader(history));
            lines = (int) count.lines().count();

            BufferedReader reader = new BufferedReader(new FileReader(history));

            if ( lines > 0) {
                for (int i = 0; i < lines; i++) {
                    if (lines > 100 && (lines - 100) > i) {
                        reader.readLine();
                        continue;
                    }
                    st = reader.readLine();
                    if (st != null || st != "")
                        Windows.text = Windows.text.concat(st+"\n");


                }
            }
            reader.close();
            count.close();
        }
    }

    static class Windows extends JFrame {

        static String text = "";

        static String name;

        static Client client;

        static JTextField field = new JTextField();
        static JTextArea area = new JTextArea();
        static JScrollPane scroll = new JScrollPane(area);


        public static void setInArea() {
            area.setText("" + text);
        }

        public void windowsName(String name) {
            setTitle(name);
        }

        public Windows(String name, Client client) {
            this.client = client;
            this.name = name;

            area.setEditable(false);
            area.setLineWrap(true);
            area.getCaret().setDot( Integer.MAX_VALUE );
            windowsName("Please LogIn");


            setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
            setBounds(200, 200, 600, 600);

            //JScrollPane jscroll = new JScrollPane(JScrollPane.VERTICAL_SCROLLBAR_ALWAYS, JScrollPane.HORIZONTAL_SCROLLBAR_AS_NEEDED);
            // jscroll.setSize(new Dimension(20,100));

            JButton btnSend = new JButton("Sent");

            field.addKeyListener(new KeyListener() {
                @Override
                public void keyTyped(KeyEvent e) {
                }

                @Override
                public void keyPressed(KeyEvent e) {
                    if (e.getExtendedKeyCode() == 10 && !(field.getText().trim().equals(""))) {
//                        text = text.concat(" " + name + " " + date() + ": " + field.getText() + "\n");
//                        area.setText("" + text);
                        sendMessege(field.getText());
                        field.setText("");
                    } else if (e.getExtendedKeyCode() == 10) {
                        field.setText("");
                    }
                }

                @Override
                public void keyReleased(KeyEvent e) {
                }

            });
            btnSend.addMouseListener(new MouseAdapter() {
                @Override
                public void mouseClicked(MouseEvent e) {
                    if (e.getClickCount() == 1 && !(field.getText().trim().equals(""))) {

//                        text = text.concat(" " + name + " " + date() + ": " + field.getText() + "\n");
//
//                        area.setText("" + text);

                        sendMessege(field.getText());

                        field.setText("");


                    } else if (e.getClickCount() == 1) {
                        field.setText("");
                    }
                }
            });

            JPanel upPanel = new JPanel();
            JPanel sendPanel = new JPanel();

            upPanel.setSize(250, 250);


            upPanel.setLayout(new BorderLayout());
            sendPanel.setLayout(new BorderLayout());
            //area.add(jscroll);
//        textField.setSize(1500,500);
            btnSend.setSize(50, 20);

            sendPanel.add(field, BorderLayout.CENTER);
            sendPanel.add(btnSend, BorderLayout.EAST);


            upPanel.add(sendPanel, BorderLayout.SOUTH);
            //       upPanel.add(btn5, BorderLayout.WEST);
//        upPanel.add(jscroll, BorderLayout.EAST);
//       upPanel.add(btn4, BorderLayout.NORTH);
            upPanel.add(scroll, BorderLayout.CENTER);

            addWindowListener(new WindowAdapter() {
                @Override
                public void windowClosing(WindowEvent e) {
                    super.windowClosing(e);
                    try {
                        client.dos.writeUTF("/quit");

                    } catch (IOException ignored) {
                    } catch (NullPointerException ignored) {
                    } finally {
                        client.closeConnection();
                    }
                }
            });


            add(upPanel);
            setVisible(true);

        }

        private void sendMessege(String text) {
            try {
                client.dos.writeUTF(text);
            } catch (Exception e) {
                System.out.println("Сообщение не отправлено");
            }
        }


        private static String date() {

            Date date = new Date();

            SimpleDateFormat ft =
                    new SimpleDateFormat("hh:mm:ss");

            return ft.format(date);


        }

    }

}
